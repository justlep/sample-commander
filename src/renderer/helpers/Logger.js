let DEFAULT_LOG_LEVEL_NAME = 'WARN',
    CONSOLE_ENABLED = (typeof console !== 'undefined'),
    NOP = function(){},
    Logger,
    getConsoleMethodOrDefault = function(consoleMethodName) {
        /*eslint no-console:0 */
        return CONSOLE_ENABLED ? (console[consoleMethodName] || console['log'] || NOP) : NOP;
    },
    getLogLevelObjectByNameOrObject = function(nameOrObject) {
        let wishedLevelObject = (typeof nameOrObject === 'string') ? Logger.LEVEL[nameOrObject] : nameOrObject;
        for (let logLevelKey in Logger.LEVEL) {
            if (!Logger.LEVEL.hasOwnProperty(logLevelKey)) {
                continue;
            }
            let levelObject = Logger.LEVEL[logLevelKey];
            if (levelObject === wishedLevelObject) {
                return levelObject;
            }
        }
        Logger.LEVEL.ERROR.fn('Invalid log level - defaulting to ' + DEFAULT_LOG_LEVEL_NAME);
        return Logger.LEVEL[DEFAULT_LOG_LEVEL_NAME];
    };

Logger = {
    LEVEL: {
        DEBUG: {prio: 10, fn: getConsoleMethodOrDefault('debug')},
        INFO:  {prio: 20, fn: getConsoleMethodOrDefault('info')},
        DEV:   {prio: 25, fn: getConsoleMethodOrDefault('log')},
        WARN:  {prio: 30, fn: getConsoleMethodOrDefault('warn')},
        ERROR: {prio: 40, fn: getConsoleMethodOrDefault('error')},
        OFF:   {prio: 50, fn: NOP}
    },
    /** @static */
    debug: NOP,
    /** @static */
    info: NOP,
    /** @static */
    dev: NOP,
    /** @static */
    warn: NOP,
    /** @static */
    error: NOP,
    /**
     * Sets a new log level. If an invalid level is given, {@link DEFAULT_LOG_LEVEL_NAME} will be used.
     * @param {Object|string} newLogLevelNameOrObject - the name of the log level OR the LEVEL.* object itself
     * @static
     */
    setLogLevel: function(newLogLevelNameOrObject) {
        let newLogLevel = getLogLevelObjectByNameOrObject(newLogLevelNameOrObject),
            newPrio = newLogLevel.prio,
            logLevel,
            logMethodName,
            isMethodEnabled;
        for (let logLevelKey in Logger.LEVEL) {
            if (Logger.LEVEL.hasOwnProperty(logLevelKey)) {
                logLevel = Logger.LEVEL[logLevelKey];
                logMethodName = logLevelKey.toLowerCase();
                isMethodEnabled = (logLevel.prio >= newPrio);
                Logger[logMethodName] = (isMethodEnabled) ? logLevel.fn : NOP;
            }
        }
    }
};

Logger.setLogLevel(DEFAULT_LOG_LEVEL_NAME);

export default Logger;
