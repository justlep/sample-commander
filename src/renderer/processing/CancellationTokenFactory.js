
const NO_TOKEN = Symbol();

/**
 * A token that can be passed to asynchronous task queues to allow signalling them
 * that task processing should be stopped (cancelled) prematurely.
 * Only ONE token from the same {@link CancellationTokenFactory} is valid (in non-cancelled state) at a time.
 */
class CancellationToken {
    /**
     * @param {CancellationTokenFactory} factory
     */
    constructor(factory) {
        this._factory = factory;
    }

    /**
     * @returns {boolean}
     */
    isCancelled() {
        return !this._factory.isTokenActive(this);
    }
    
    cancel() {
        this._factory.cancelToken(this);
    }
}

/**
 * A Factory producing {@link CancellationToken} instances.
 * Only ONE token is valid (i.e. in non-cancelled state) at a time.
 * Generating a new token automatically cancels the previous one.
 */
export class CancellationTokenFactory {
    
    constructor() {
        this._activeToken = NO_TOKEN;
    }

    /**
     * Creates a new token, implicitly cancelling the previously active token.
     * @returns {CancellationToken}
     */
    newToken() {
        this._activeToken = new CancellationToken(this);
        return this._activeToken;
    }

    /**
     * Invalidates the currently active token. 
     */
    cancelCurrentToken() {
        this._activeToken = NO_TOKEN;
    }

    /**
     * Invalidates the given token if it is currently active.
     * @param {CancellationToken} token
     */
    cancelToken(token) {
        this.isTokenActive(token) && this.cancelCurrentToken();
    }
    
    /**
     * @param {*} token
     * @returns {boolean}
     */
    isTokenActive(token) {
        return token === this._activeToken;
    }
}
