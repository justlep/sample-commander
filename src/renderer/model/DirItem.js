import nodePath from 'path'
import {assertDirExists} from '@/helpers/fileHelper'
import readdirp from 'readdirp';
import Logger from '@/helpers/Logger'
import {NON_REACTIVE_PROPS} from '@/constants'

const SEP_REGEX = nodePath.sep === '\\' ? /\\/g : /\//g;

let _dirIds = 1,
    nextId = () => (++_dirIds).toString(36);


export default class DirItem {

    /**
     * @param {string} path
     * @param {number} [limit] - the maximum # of dirs to load (if 0, no limit is used)
     * @param {CancellationToken} cancellationToken - checked frequently during the tree walk. 
     *                                                When isCancelled() is true, tree-walk will be aborted after the currently processed file.
     * @param {boolean} [includeRootPath] - if true, the start path is returned as the first item, followed by its descendants                              
     * @return {Promise<{dirItems: DirItem[], isLimitExceeded: Boolean}>}
     */
    static loadRecursively({path, limit = Number.MAX_SAFE_INTEGER, cancellationToken, includeRootPath = false}) {
        if (isNaN(limit)) {
            return Promise.reject('Invalid limit ' + limit);
        }
        const effectiveLimit = limit || Number.MAX_SAFE_INTEGER;
        return assertDirExists(path).then((validatedRootPath) => new Promise((resolve, reject) => {
            let dirItems = includeRootPath ? [new DirItem(validatedRootPath)] : [],
                stream = readdirp(path, {
                    type: 'directories',
                    depth: Number.MAX_SAFE_INTEGER,
                    // (!) readdirp crashes on *.asar files when used within Electron apps and asar enabled
                    //     because Electron "augments" Node's native fs package to treat .asar files like directories, 
                    //     but it breaks readdirp. 
                    //     
                    //     Possible solutions: 
                    //          (1) set process.noAsar for renderer process => NO, breaks the app 
                    //          (2) exclude .asar via filter => works  
                    directoryFilter: ['!*.asar']  
                }),
                isLimitExceeded = false,
                finishedCallback = () => {
                    DirItem.sortList(dirItems);
                    resolve({dirItems, isLimitExceeded});
                },
                onData = ({fullPath}) => {
                    isLimitExceeded = dirItems.length >= effectiveLimit;
                    if (isLimitExceeded || (cancellationToken && cancellationToken.isCancelled())) {
                        Logger.debug('destroy dir stream (cancelled or limit exceeded)');
                        stream.pause();
                        stream.destroy();
                        return;
                    }
                    dirItems.push(new DirItem(fullPath));
                };

            stream
                .on('data', onData)
                .on('warn', err => Logger.warn(err))
                .on('error', err => {
                    Logger.error(err);
                    reject(err);
                })
                //.on('end', finishedCallback)
                .on('close', () => {
                    Logger.debug('close DirItem stream');
                    finishedCallback();
                })
                .on('end', () => {
                    Logger.debug('end DirItem stream');
                    finishedCallback();
                });
            })
        );
    }

    /**
     * @param {FileItem[]} dirItems
     * @return {FileItem[]} 
     */
    static sortList(dirItems) {
        return dirItems.sort((a, b) => a.path.localeCompare(b.path));
    }

    /**
     * @param {string} path
     */
    constructor(path) {
        const resolvedPath = nodePath.resolve(path);
        
        this[NON_REACTIVE_PROPS] = {
            path: resolvedPath,
            name: nodePath.basename(resolvedPath),
            depth: path.match(SEP_REGEX).length,
            id: nextId() // TODO is there still a use for this?
        };
    }
    
    get path() { return this[NON_REACTIVE_PROPS].path }
    get name() { return this[NON_REACTIVE_PROPS].name }
    get depth() { return this[NON_REACTIVE_PROPS].depth }
    get id() { return this[NON_REACTIVE_PROPS].id }
}
