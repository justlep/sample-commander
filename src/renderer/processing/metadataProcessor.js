import {execFile} from 'child_process';
import Logger from '@/helpers/Logger'
import {format} from 'util';

/**
 * Determines metadata (e.g. sample rate) for this file using the ffprobe executable.
 * If the metadata is in the cache, that cached version is returned.
 *
 * @param {FileItem} fileItem
 * @param {string} ffprobeExecutablePath
 * @returns {Promise<{duration: String, sampleRate: Number}>}
 */
export async function getMetadataForFileItem(fileItem, ffprobeExecutablePath) {
    if (!ffprobeExecutablePath) {
        throw new Error('Invalid ffprobeExecutablePath');
    }

    let metadataJson;
    
    try {
        let metadataJsonString = await new Promise((resolve, reject) => {
                Logger.dev('Running ffprobe for metadata of %s', fileItem.path);
    
                execFile(ffprobeExecutablePath, [
                    '-hide_banner',
                    '-show_streams',
                    '-select_streams', 'a:0',
                    // '-sexagesimal',  // <-- for duration getting formatted like "0:09:39.396984" 
                    '-show_entries', 'stream=duration,duration_ts,sample_rate,time_base',
                    '-print_format', 'json',
                    fileItem.path
                ], {
                    timeout: 3000,
                    shell: false,
                    encoding: 'utf8',
                    stdio: ['pipe', 'pipe', 'ignore']
                }, (err, stdout, stderr) => {
                    if (err) {
                        reject(format('ffprobe failed to load metadata for %s, error: %s', fileItem.path, stderr));
                    } else {
                        resolve('' + stdout);
                    }
                });
            }),
            json = JSON.parse(metadataJsonString),
            firstStream = json && json.streams[0],
            sampleRate,
            duration,
            durationInMins;
        
        if (firstStream) {
            sampleRate = firstStream && parseInt(firstStream['sample_rate'], 10);
            
            let durationInSecs = parseFloat(firstStream.duration);

            // Logger.dev(firstStream);
            
            if (!isNaN(durationInSecs)) {
                // duration is in minutes
                let hours = (durationInSecs / 3600) << 0, 
                    mins = ((durationInSecs - (hours * 3600)) / 60) << 0, 
                    secs = (durationInSecs % 60) << 0,
                    parts = [];

                if (durationInSecs >= 1) {
                    if (hours) {
                        parts.push(hours + 'h');
                    }
                    if (mins) {
                        parts.push(mins + 'm');
                    }
                    if (secs) {
                        parts.push(secs + 's');
                    }
                } else {
                    parts.push(((durationInSecs * 1000) << 0) + 'ms');
                }
                duration = parts.join(' ');
                durationInMins = durationInSecs / 60;
            }
        }

        metadataJson = {
            sampleRate: sampleRate || null, 
            duration,
            durationInMins
        };
        
    } catch (err) {
        Logger.error('Failed to process metadata: %o', err);
        throw new Error(format('Failed to process metadata JSON  for %s, error: %s', fileItem.path, err));
    }
    
    // Logger.dev('Finished metadata for %s -> %o', fileItem.filename, metadataJson);
    return metadataJson;
}
