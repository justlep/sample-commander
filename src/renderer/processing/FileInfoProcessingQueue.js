import {EventEmitter} from 'events'
import Logger from '@/helpers/Logger'
import {priorityQueue, memoize, asyncify} from 'async'
import {getSpectrogramForFileItem} from './spectrogramProcessor'
import {getMetadataForFileItem} from './metadataProcessor'

const DEFAULT_CONCURRENCY = 1;

const STATUS_MAP_SYMBOL = Symbol('PROCESSING_STATUS_BY_FILE_ITEM_ID');

const STATUS_QUEUED = 1;
const STATUS_PROCESSING = 2;
const STATUS_ERROR = 4;

export const EVENT_UNFINISHED_METADATA_COUNT = 'unfinished-metadata-count';
export const EVENT_UNFINISHED_SPECTROGRAM_COUNT = 'unfinished-spectrogram-count';

let _instance;


/**
 * @param {AsyncFunction} asyncFileItemProcessingFn - an `AsyncFunction` in terminology of the async library, NOT an ES2017 async function. 
 *                        e.g. function(fileItem, callback) { .. callback(null, processedData); }
 * @param {function} onProcessedDataReceived, e.g. function(processedData){..}
 * @return {QueueObject}
 * @private
 */
function _createFileItemQueue(asyncFileItemProcessingFn, onProcessedDataReceived) {
    let queue = priorityQueue(function(fileItem, callback) {
        _setFileItemQueueStatus(fileItem, queue, STATUS_PROCESSING);
        asyncFileItemProcessingFn(fileItem, function(err, processedData) {
            _setFileItemQueueStatus(fileItem, queue, null);
            if (err) {
                return callback(err);
            }
            onProcessedDataReceived(fileItem, processedData);
            callback(null, processedData);
        });
    }, DEFAULT_CONCURRENCY);

    queue[STATUS_MAP_SYMBOL] = {};
    queue.error = ((err, fileItem) => {
        Logger.warn('Error while processing %s: %s', fileItem.toString(), err);
        _setFileItemQueueStatus(fileItem, queue, STATUS_ERROR);
    });

    return queue;
}

/**
 * @param {QueueObject} queue
 * @param {number} concurrency
 */
function _setQueueConcurrency(queue, concurrency) {
    queue.concurrency = concurrency;
    queue.buffer = Math.max(0, concurrency - 1);
    if (!queue.paused) {
        // TODO not sure if changing the concurrency really requires pause/resume, doesnt hurt though. check later.
        queue.pause();
        queue.resume();
    }
}

/**
 * @param {FileItem} fileItem
 * @param {QueueObject} queue
 * @returns {number} - one of the STATUS_* constants
 * @private
 */
function _getFileItemQueueStatus(fileItem, queue) {
    return queue[STATUS_MAP_SYMBOL][fileItem.id];
}

/**
 * @param {FileItem} fileItem
 * @param {QueueObject} queue
 * @param {number|null} status
 * @private
 */
function _setFileItemQueueStatus(fileItem, queue, status) {
    if (!status) {
        delete queue[STATUS_MAP_SYMBOL][fileItem.id];
    } else {
        queue[STATUS_MAP_SYMBOL][fileItem.id] = status;
    }
}

function _clearAllFileItemQueueStatus(queue) {
    queue[STATUS_MAP_SYMBOL] = {};
}


function _flushQueue(queue) {
    queue.remove(() => true);
    if (queue.idle()) {
        _clearAllFileItemQueueStatus(queue);
    }
}

/**
 * @param {FileItem} fileItem
 * @param {string} duration
 * @param {number} durationInMins
 * @param {number} sampleRate
 */
function _applyMetadataProcessorResultToFileItem(fileItem, {duration, durationInMins, sampleRate}) {
    fileItem.metadata.duration = duration;
    fileItem.metadata.sampleRate = sampleRate;
    fileItem.metadata.durationInMins = durationInMins;
}

/**
 * @param {FileItem} fileItem
 * @param {string} spectrogramPath
 */
function _applySpectrogramProcessorResultToFileItem(fileItem, spectrogramPath) {
    fileItem.spectrogram = spectrogramPath;
}

/**
 * @param {FileItem} fileItem
 * @param {Object<string, *>} idBasedCache
 * @param {function} applyFn - one of the _apply*ToFileItem() methods above
 * @return {boolean} - true if a cached value was found & applied
 */
function _applyCachedProcessorResultIfExists(fileItem, idBasedCache, applyFn) {
    let cacheEntry = idBasedCache[fileItem.id],
        cachedProcessorResult = cacheEntry && cacheEntry[1],
        isApplicable = !!cachedProcessorResult;

    if (isApplicable) {
        applyFn(fileItem, cachedProcessorResult);
    }
    return isApplicable;
}

/**
 * Represent a processing managing two queues for FFmpeg-based gathering of information for lists of 
 */
export class FileInfoProcessingQueue extends EventEmitter {
    
    /** 
     * @return {FileInfoProcessingQueue} 
     */
    static getInstance() {
        return _instance || (_instance = new FileInfoProcessingQueue());
    }

    /**
     * @param {string} [ffmpegExecutablePath]
     * @param {string} [ffprobeExecutablePath]
     * @param {string} [linearSpectrogramDir]
     * @param {number} [metadataConcurrency]
     * @param {number} [spectrogramConcurrency]
     * @param {boolean} [smallFilesFirst]
     * @param {number} [spectrogramSourceFileSizeLimitInMb]
     * @param {number} [spectrogramSourceFileDurationLimitInMins]
     * @returns {FileInfoProcessingQueue}
     */
    configure({ffmpegExecutablePath, ffprobeExecutablePath, linearSpectrogramDir, metadataConcurrency, spectrogramConcurrency, 
                  smallFilesFirst, spectrogramSourceFileSizeLimitInMb, spectrogramSourceFileDurationLimitInMins}) {
        const logPattern = '[FileInfoProcessingQueue] setting %s = %s';
        if (ffmpegExecutablePath) {
            Logger.dev(logPattern, 'ffmpegExecutablePath', ffmpegExecutablePath);
            this._ffmpegExecutablePath = ffmpegExecutablePath;
        }
        if (ffprobeExecutablePath) {
            Logger.dev(logPattern, 'ffprobeExecutablePath', ffprobeExecutablePath);
            this._ffprobeExecutablePath = ffprobeExecutablePath;   
        }
        if (linearSpectrogramDir) {
            Logger.dev(logPattern, 'linearSpectrogramDir', linearSpectrogramDir);
            this._linearSpectrogramDir = linearSpectrogramDir;   
        }
        if (metadataConcurrency && typeof metadataConcurrency === 'number') {
            Logger.dev(logPattern, 'metadataConcurrency', metadataConcurrency);
            _setQueueConcurrency(this._metadataQueue, metadataConcurrency);
        }
        if (spectrogramConcurrency && typeof spectrogramConcurrency === 'number') {
            Logger.dev(logPattern, 'spectrogramConcurrency', spectrogramConcurrency);
            _setQueueConcurrency(this._spectrogramQueue, spectrogramConcurrency);
        }
        if (typeof smallFilesFirst === 'boolean') {
            Logger.dev(logPattern, 'smallFilesFirst', smallFilesFirst);
            this._smallFilesFirst = smallFilesFirst;
        }
        if (typeof spectrogramSourceFileSizeLimitInMb === 'number') {
            Logger.dev(logPattern, 'spectrogramSourceFileSizeLimitInMb', spectrogramSourceFileSizeLimitInMb);
            this._spectrogramSourceFileSizeLimitInBytes = (spectrogramSourceFileSizeLimitInMb * 1024 * 1024) || Number.MAX_SAFE_INTEGER;
        }
        if (typeof spectrogramSourceFileDurationLimitInMins === 'number') {
            Logger.dev(logPattern, 'spectrogramSourceFileDurationLimitInMins', spectrogramSourceFileDurationLimitInMins);
            this._spectrogramSourceFileDurationLimitInMins = spectrogramSourceFileDurationLimitInMins || Number.MAX_SAFE_INTEGER;
        }
        return this;
    }
    
    flushAllQueues() {
        [this._spectrogramQueue, this._metadataQueue].forEach(queue => {
            _flushQueue(queue);
            queue.resume();
        });
    }

    /**
     * @param {QueueObject} queue
     * @param {string} eventName - the event to be emitted with the # of the queue's unfinished jobs (only emitting upon changes)
     * @returns {function(): boolean} - always returns true 
     * @private
     */
    _createUnfinishedJobsEventEmittingFn(queue, eventName) {
        let prevCount = 0;
        return () => {
            let countNow = queue.length() + queue.running();
            if (countNow !== prevCount) {
                prevCount = countNow;
                this.emit(eventName, countNow);
            }
            return true;
        }
    }

    constructor() {
        super();
        if (_instance) {
            throw new Error('Use FileInfoProcessor.getInstance()');
        }

        /** @type {string|null} */
        this._ffmpegExecutablePath = '';
        /** @type {string|null} */
        this._ffprobeExecutablePath = '';

        /** @type {string} */
        this._linearSpectrogramDir = '';
        
        this._smallFilesFirst = true;
        
        /** @type {number} */
        this._spectrogramSourceFileSizeLimitInBytes = Number.MAX_SAFE_INTEGER;
        
        /** @type {number} */
        this._spectrogramSourceFileDurationLimitInMins = Number.MAX_SAFE_INTEGER;
        
        const self = this;
        
        const cachingMetadataWorkerFn = memoize(
            asyncify(fileItem => getMetadataForFileItem(fileItem, self._ffprobeExecutablePath)),
            fileItem => fileItem.id
        );
        
        const cachingSpectrogramWorkerFn = memoize(
            asyncify(fileItem => getSpectrogramForFileItem(fileItem, self._ffmpegExecutablePath, self._linearSpectrogramDir)),
            fileItem => fileItem.id
        );
        
        this._metadataQueue = _createFileItemQueue(cachingMetadataWorkerFn, _applyMetadataProcessorResultToFileItem);
        
        this._spectrogramQueue = _createFileItemQueue(cachingSpectrogramWorkerFn, _applySpectrogramProcessorResultToFileItem);

        /** @type {Object<string, {duration: string, sampleRate: number, durationInMins: number}>} */
        this._cachedMetadataByFileItemId = cachingMetadataWorkerFn.memo;
        /** @type {Object<string, string>} */    
        this._cachedSpectrogramByFileItemId = cachingSpectrogramWorkerFn.memo;

        const _emitUnfinishedMetadata = this._createUnfinishedJobsEventEmittingFn(this._metadataQueue, EVENT_UNFINISHED_METADATA_COUNT);
        const _emitUnfinishedSpectrograms = this._createUnfinishedJobsEventEmittingFn(this._spectrogramQueue, EVENT_UNFINISHED_SPECTROGRAM_COUNT);
        
        const _toggleSpectroQueueWhileMetadataProcessing = () => {
            let isMetaIdle = this._metadataQueue.idle(),
                isSpectroPaused = this._spectrogramQueue.paused;    
            
            if (isMetaIdle && isSpectroPaused) {
                Logger.dev('Unpausing spectrogram queue');
                this._spectrogramQueue.resume();
            } else if (!isMetaIdle && !isSpectroPaused) {
                Logger.dev('Pausing spectrogram queue');
                this._spectrogramQueue.pause();
            }
        };
        
        // metadata queue status changes...
        
        this._metadataQueue.saturated = () => _emitUnfinishedMetadata() && _toggleSpectroQueueWhileMetadataProcessing(); 
        this._metadataQueue.unsaturated = () => _emitUnfinishedMetadata() && _toggleSpectroQueueWhileMetadataProcessing();
        this._metadataQueue.empty = () => {
            Logger.dev('Metadata task queue is empty');
            _toggleSpectroQueueWhileMetadataProcessing();
            _emitUnfinishedMetadata();
        };
        this._metadataQueue.drain = () => {
            Logger.dev('Metadata task queue is drained, cache: %o', this._cachedMetadataByFileItemId);
            this._spectrogramQueue.resume();
            _emitUnfinishedMetadata();
            _clearAllFileItemQueueStatus(this._metadataQueue);
        };

        // spectrogram queue status changes...
        
        this._spectrogramQueue.saturated = _emitUnfinishedSpectrograms;
        this._spectrogramQueue.unsaturated = _emitUnfinishedSpectrograms;
        this._spectrogramQueue.empty = () => {
            Logger.dev('Spectrogram task queue is empty');
            _emitUnfinishedSpectrograms();
        };
        this._spectrogramQueue.drain = () => {
            _emitUnfinishedSpectrograms();
            _clearAllFileItemQueueStatus(this._spectrogramQueue);
            Logger.dev('Spectrogram task queue is drained, cache: %o', this._cachedSpectrogramByFileItemId);
        };

        
        // TODO persist/re-import queues' cache data
    }
    
    /**
     * @param {FileItem} fileItem
     * @param {QueueObject} queue
     * @private
     */
    _scheduleProcessing(fileItem, queue) {
        let purpose = queue === this._metadataQueue ? 'metadata' : 'spectrogram';
        if (!_getFileItemQueueStatus(fileItem, queue)) {
            Logger.dev('Scheduling %s processing of %s', purpose, fileItem.path);
            let priority = this._smallFilesFirst ? fileItem.filesize : 1;
            _setFileItemQueueStatus(fileItem, queue, STATUS_QUEUED);
            queue.push(fileItem, priority);
        } else {
            Logger.dev('Skipped re-scheduling %s processing of %s', purpose, fileItem.path);
        }
    }

    get metadataConcurrency() { return this._metadataQueue.concurrency }
    get spectrogramConcurrency() { return this._spectrogramQueue.concurrency }

    /**
     * Schedules the metadata processing for the given file item if it's not already loaded or scheduled
     * @param {FileItem} fileItem
     */
    scheduleMetadataProcessing(fileItem) {
        if (!fileItem.hasMetadata()) {
            if (_applyCachedProcessorResultIfExists(fileItem, this._cachedMetadataByFileItemId, _applyMetadataProcessorResultToFileItem)) {
                Logger.dev('Used cached value for metadata of %s', fileItem.path);
                return;
            }
            this._scheduleProcessing(fileItem, this._metadataQueue);
        } else {
            Logger.dev('Skipped scheduling %s for metadata processing', fileItem);
        }
    }

    /**
     * Schedules the spectrogram processing for the given file item if it's not already loaded or scheduled
     * @param {FileItem} fileItem
     */
    scheduleSpectrogramProcessing(fileItem) {
        if (!fileItem.supportsSpectrograms) {
            return;
        }
        if (!fileItem.spectrogram) {
            if (_applyCachedProcessorResultIfExists(fileItem, this._cachedSpectrogramByFileItemId, _applySpectrogramProcessorResultToFileItem)) {
                Logger.dev('Used cached value for spectrogram of %s', fileItem.path);
                return;
            }
            if (fileItem.filesize > this._spectrogramSourceFileSizeLimitInBytes) {
                Logger.dev('Skipped scheduling too big %s for spectrogram processing', fileItem);
                return;
            }
            if ((fileItem.metadata.durationInMins || 0) > this._spectrogramSourceFileDurationLimitInMins) {
                Logger.dev('Skipped scheduling too long %s for spectrogram processing', fileItem);
                return;
            }
            if (fileItem.metadata.sampleRate === undefined) {
                this.scheduleMetadataProcessing(fileItem);
            }
            this._scheduleProcessing(fileItem, this._spectrogramQueue);
        } else {
            Logger.dev('Skipped scheduling %s for spectrogram processing', fileItem);
        }
    }

    /**
     * Removes the cached metadata & spectrogram from the cache (if any).
     * @param {FileItem|string} fileItemOrId
     */
    flushCachesForFileItem(fileItemOrId) {
        let id = (typeof fileItemOrId === 'string') ? fileItemOrId : fileItemOrId.id; 
        delete this._cachedMetadataByFileItemId[id];
        this.flushSpectrogramCacheForFileItem(id);
    }

    /**
     * Flush the spectrogram path cache for a given file item
     * @param {FileItem|string} fileItemOrId
     */
    flushSpectrogramCacheForFileItem(fileItemOrId) {
        let id = (typeof fileItemOrId === 'string') ? fileItemOrId : fileItemOrId.id;
        Logger.warn('Clearing spectrogram cache for ' + fileItemOrId);
        delete this._cachedSpectrogramByFileItemId[id];
    }
    
    flushAllCaches() {
        Object.keys(this._cachedMetadataByFileItemId).forEach(key => delete this._cachedMetadataByFileItemId[key]);
        Object.keys(this._cachedSpectrogramByFileItemId).forEach(key => delete this._cachedSpectrogramByFileItemId[key]);
        Logger.warn('Flushed all processing caches');
    }
}
