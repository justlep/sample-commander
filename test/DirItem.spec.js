import DirItem from '@/model/DirItem'
import Logger from '@/helpers/Logger';
import {CancellationTokenFactory} from '@/processing/CancellationTokenFactory';
import path from 'path'

const TEST_DIR_PATH = path.normalize(__dirname);

describe('DirItem', () => {

    const tokenFactory = new CancellationTokenFactory();

    beforeAll(() => {
        Logger.setLogLevel(Logger.LEVEL.OFF)
    });

    it('can scan for folders, excluding the start path', async () => {
        let cancellationToken = tokenFactory.newToken(),
            {dirItems, isLimitExceeded} = await DirItem.loadRecursively({path: TEST_DIR_PATH, cancellationToken});

        expect(isLimitExceeded).toBe(false);
        expect(Array.isArray(dirItems)).toBe(true);
        expect(dirItems.find(it => it.path === TEST_DIR_PATH)).toBeFalsy(); // root not included by default
        expect(dirItems.map(it => it.name)).toMatchSnapshot();
    });

    it('can scan for folders, including the start path', async () => {
        let cancellationToken = tokenFactory.newToken(),
            {dirItems, isLimitExceeded} = await DirItem.loadRecursively({path: TEST_DIR_PATH, cancellationToken, includeRootPath: true});

        expect(isLimitExceeded).toBe(false);
        expect(Array.isArray(dirItems)).toBe(true);
        expect(dirItems.find(it => it.path === TEST_DIR_PATH)).toBeInstanceOf(DirItem); // root included
        expect(dirItems.map(it => it.name)).toMatchSnapshot();
    });

    it('can scan for folders with a limit', async () => {
        let cancellationToken = tokenFactory.newToken(),
            {dirItems, isLimitExceeded} = await DirItem.loadRecursively({path: TEST_DIR_PATH, cancellationToken, includeRootPath: true});

        expect(isLimitExceeded).toBe(false);
        const totalUnlimited = dirItems.length;
        expect(totalUnlimited).toBeGreaterThan(2);

        const LIMIT = totalUnlimited - 1;
       
        // limit < # of existing paths
        let retObj = await DirItem.loadRecursively({path: TEST_DIR_PATH, cancellationToken, includeRootPath: true, limit: LIMIT});
        expect(retObj.dirItems.length).toBe(LIMIT);
        expect(retObj.isLimitExceeded).toBe(true);

        // limit === # of existing paths
        retObj = await DirItem.loadRecursively({path: TEST_DIR_PATH, cancellationToken, includeRootPath: true, limit: totalUnlimited});
        expect(retObj.dirItems.length).toBe(totalUnlimited);
        expect(retObj.isLimitExceeded).toBe(false); // limit is reached, but not exceeded
    });

    it('can be interrupted scanning for folders via cancellationToken', async () => {
        let cancellationToken = tokenFactory.newToken();
        cancellationToken.cancel();
        let {dirItems, isLimitExceeded} = await DirItem.loadRecursively({path: TEST_DIR_PATH, cancellationToken});

        expect(isLimitExceeded).toBe(false);
        expect(Array.isArray(dirItems)).toBe(true);
        expect(dirItems.length).toBe(0);
    });
    
    it('determines the depth of each folder', async () => {
        let cancellationToken = tokenFactory.newToken(),
            {dirItems, isLimitExceeded} = await DirItem.loadRecursively({path: TEST_DIR_PATH, cancellationToken, includeRootPath: true});
        
        expect(dirItems.length).toBeGreaterThan(1);
        const rootDepth = dirItems[0].depth;
        expect(rootDepth).toBeGreaterThan(1);
        
        for (let i = 1; i < dirItems.length; i++) {
            expect(dirItems[i].depth).toBeGreaterThan(rootDepth);
        }
        
        expect(dirItems.find(it => it.name === 'subdir').depth).toBe(rootDepth + 2);
    });
});
