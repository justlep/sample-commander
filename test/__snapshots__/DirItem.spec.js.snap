// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirItem can scan for folders, excluding the start path 1`] = `
Array [
  "__snapshots__",
  "test-files",
  "test-files-generated",
  "subdir",
]
`;

exports[`DirItem can scan for folders, including the start path 1`] = `
Array [
  "test",
  "__snapshots__",
  "test-files",
  "test-files-generated",
  "subdir",
]
`;
